From 7ed98ba9fda3dfcd72d277dfd4f4db53eaa67cc8 Mon Sep 17 00:00:00 2001
From: Sun Yuechi <sunyuechi@iscas.ac.cn>
Date: Tue, 2 Jun 2026 13:07:31 +0800
Subject: [PATCH] test: add RISC-V architecture probe tests

Add RISC-V ZBC/ZVBC extension checks to test_arch.cc.

Normalize riscv.c probe results to 0/1,
so the test comparison works consistently with other architectures.

Signed-off-by: Sun Yuechi <sunyuechi@iscas.ac.cn>
---
 src/arch/riscv.c      |  9 ++++++---
 src/test/test_arch.cc | 28 ++++++++++++++++++++++++++++
 2 files changed, 34 insertions(+), 3 deletions(-)

diff --git a/src/arch/riscv.c b/src/arch/riscv.c
index 5dc54acfd86..b17a0f87bbc 100644
--- a/src/arch/riscv.c
+++ b/src/arch/riscv.c
@@ -35,8 +35,11 @@ void ceph_arch_riscv_probe(void)
 
     if (do_hwprobe(pairs, 1) == 0) {
         unsigned long long ext = pairs[0].value;
-        ceph_arch_riscv_rvv  = (ext & RISCV_HWPROBE_IMA_V);
-        ceph_arch_riscv_zbc  = (ext & RISCV_HWPROBE_EXT_ZBC);
-        ceph_arch_riscv_zvbc = (ext & RISCV_HWPROBE_EXT_ZVBC);
+        ceph_arch_riscv_rvv =
+            (ext & RISCV_HWPROBE_IMA_V) == RISCV_HWPROBE_IMA_V;
+        ceph_arch_riscv_zbc =
+            (ext & RISCV_HWPROBE_EXT_ZBC) == RISCV_HWPROBE_EXT_ZBC;
+        ceph_arch_riscv_zvbc =
+            (ext & RISCV_HWPROBE_EXT_ZVBC) == RISCV_HWPROBE_EXT_ZVBC;
     }
 }
diff --git a/src/test/test_arch.cc b/src/test/test_arch.cc
index b53d84fdd7d..6be6c07d840 100644
--- a/src/test/test_arch.cc
+++ b/src/test/test_arch.cc
@@ -21,6 +21,7 @@
 #include "arch/intel.h"
 #include "arch/arm.h"
 #include "arch/ppc.h"
+#include "arch/riscv.h"
 #include "global/global_context.h"
 #include "gtest/gtest.h"
 
@@ -97,6 +98,33 @@ TEST(Arch, all)
 #endif
 
 #endif
+
+#if __riscv && __linux__
+  char isa[FLAGS_SIZE];
+  FILE *f_isa = popen("grep '^isa[	 ]*:' /proc/cpuinfo | head -1", "r");
+  if (f_isa == NULL || fgets(isa, FLAGS_SIZE - 1, f_isa) == NULL) {
+    if (f_isa)
+      pclose(f_isa);
+    return;
+  }
+  pclose(f_isa);
+  // Replace '_' with ' ' so we can match extensions with strstr(" ext ")
+  for (char *p = isa; *p; p++) {
+    if (*p == '_')
+      *p = ' ';
+  }
+  size_t isa_len = strlen(isa);
+  if (isa_len > 0 && isa[isa_len - 1] == '\n')
+    isa[isa_len - 1] = ' ';
+
+  int expected;
+
+  expected = strstr(isa, " zbc ") ? 1 : 0;
+  EXPECT_EQ(expected, ceph_arch_riscv_zbc);
+
+  expected = strstr(isa, " zvbc ") ? 1 : 0;
+  EXPECT_EQ(expected, ceph_arch_riscv_zvbc);
+#endif
 }
 
 
-- 
2.55.0

